[SIGSRE-58] add guide on picking good SLOs (#25)
operate-first/SRE: picking_good_slos.md

An SLO is used to quantify customer experience.
Furthermore, SLOs enable us to avoid false positives in alerting and help us focus on symptom-based alerting.
SLOs can vary from organization to organization. Consequently, there is no single formula that we could use to define the appropriate reliability target. 
In this article we will share some tips for picking a good SLO.

So, what should we consider when picking the SLO for a service?

It's important to understand the outcomes of setting an SLO. These outcomes differ depending on multiple factors, including both the maturity of the service and the whole organization overall.
To understand the different SLO settings, let's explore some use-cases that appear at either end of SLO adoption.

## Case 1 - New Service

For this case, imagine a mature SRE organization. In this organization, we have identified stakeholders spanning SLO [personas](./ADRs/RH/SIG-SRE/ADR-00002%20Personas%20related%20to%20Managed%20Services.md).
Each stakeholder understands the outcomes of both meeting or missing an SLO. 

So with that in mind, what should be our service level objective? 

> The fact that this case has a new service brings forward the opportunity to use an SLO to steer development efforts before reaching production. 

Reviewer: s/The fact that this case has a new service brings forward/A new service presents

> Ideally, a new service in production is both featureful and reliable. Product managers responsible for the service could use an SLO to define the production readiness by its limited and general availability or other aspects of the service. The availability of a service typically has deadlines, and an SLO can assist in making a clear decision about the risk for a service.

Reviewer: confused a bit by the words availability and deadline here.  Did you mean "The production readiness of a service typically has go-to-market timelines from the business.  SLOs can assist in making clear and quantifiable decisions about the service's readiness.

Author: well-said Jeremy!

> Do we focus our remaining time on making the service more reliable or adding a feature that customer X wanted?

Reviewer: s/remaining//

> The answer to the above is your decision to make - SLO is the tool used to remove subjectivity about components of the decision. 

Reviewer: Please add a link to the RACI doc here. https://github.com/operate-first/sre/blob/main/ADRs/RH/SIG-SRE/ADR-00006%20SLO%20RACI%20Chart.md

Since we have exceeded our SLO because of a third-party service, should we stop releasing new features?

When creating an SLO, external dependencies should be considered, and mitigation strategies should be employed to handle service disruptions from those external dependencies. 
Once your organization is convinced of the benefits that SLOs bring across multiple teams, retrofitting this into a mature service has some cultural considerations. 

First, all stakeholders have to agree on the outcomes of missing an SLO [burning the error budget](./ADRs/RH/SIG-SRE/ADR-00007-Error-Budget-Policy.md). 
This decision may take time for engineering teams to digest the idea of slowing feature work and focusing on reliability issues. 
The goal here should be the balance of direct impact on current processes vs introducing new processes. 

There are two main paths this could be navigated, and each should consider if there is pre-existing data that provides a baseline of current service levels. 
* Set a high SLO to start with and work towards it. What is high? 99.9%? This decision depends on if you can approximate the current service levels being delivered. 

> * Set a low SLO to be more tight interally and catch potential issues earlier. However, tight SLOs lead to higher alerting frequency. 

Reviewer: Our SLO Lifecycle document talks about a "Permissive SLO" concept to help fine tune an SLO and introduce new ones.  https://github.com/operate-first/sre/blob/main/ADRs/RH/SIG-SRE/ADR-00005%20SLO%20Lifecycle.md#permissive-slo-phase=

Reviewer: Yes I think so.

# How to pick a good SLO

> A Service Level Objective (SLO) is a measurable, specific goal for a given service's reliability level of service over time, based on one or more SLIs.

Reviewer 2: To be consistent, I would suggest the expansion of the SLI acronym (like you did for SLOs at the beginning of this sentence).

Also, I don't think it would hurt to actually say what an SLI is even though you are referencing the "Picking good SLIs" document later on.

How about adding something like this to the end of this sentence:

The SLI is a measure of "what-you-want" (e.g. x should be true) and the SLO is the measure of "how-much-you-want-it" (e.g. y% of the time). Ideally, the SLOs/SLIs are reflecting of some aspect of the service which a customer would consider necessary to the satisfactory usage of that service. For example, as a customer, it is really important for me to be able to get an answer back for my search within 5 seconds (what I want) at least 99% of the time (how much I want it). These are called "customer-centric" SLOs and the reflect a customer's view of the "reliability" they expect for that service.

To learn more about SLIs, check out our article on how to [pick good SLIs](./picking_good_slis.md).

An SLO is used to quantify customer experience.

> Furthermore, SLOs enable us to avoid false positives in alerting and help us focus on symptom-based alerting.

Reviewer 3: ```suggestion
Furthermore, SLOs enable service owners to avoid false positives in alerting and help focus on symptom-based alerting.
```

Removing ambiguousness because who is "us" referencing? The authors of this ADR, or the reader?

"service owners" could even link to the persona, eg `https://github.com/operate-first/sre/blob/main/ADRs/RH/SIG-SRE/ADR-00002%20Personas%20related%20to%20Managed%20Services.md#service-owner`.

> symptom-based alerting

I believe this is a new term in the context of the Operate First documents? If it isn't new, this would be great text to hyperlink to its usage/definition/explanation. Otherwise, do not forget to define the term and discuss why it's something to be focusing on.

SLOs can vary from organization to organization. Consequently, there is no single formula that we could use to define the appropriate reliability target. 
In this article we will share some tips for picking a good SLO.

> So, what should we consider when picking the SLO for a service?

Reviewer 3: ```suggestion
So, what should be considered when picking the SLO for a service?
```

Removing ambiguity.

> 

Reviewer 2: How would you feel about adding a short paragraph here along the lines of:

As stated previously, the SLO is based on good SLIs, and these should reflect the aspects of the service of most importance to the customer. All the considerations following are nuances of this underlying theme.

It's important to understand the outcomes of setting an SLO. These outcomes differ depending on multiple factors, including both the maturity of the service and the whole organization overall.
To understand the different SLO settings, let's explore some use-cases that appear at either end of SLO adoption.

## Case 1 - New Service

For this case, imagine a mature SRE organization. In this organization, we have identified stakeholders spanning SLO [personas](./ADRs/RH/SIG-SRE/ADR-00002%20Personas%20related%20to%20Managed%20Services.md).
Each stakeholder understands the outcomes of both meeting or missing an SLO. 

> So with that in mind, what should be our service level objective? 

Reviewer 3: ```suggestion
So with that in mind, what should the service level objective be?
```

Ambiguity.

A new service presents the opportunity to use an SLO to steer development efforts before reaching production. 

> Ideally, a new service in production is both featureful and reliable. Product managers responsible for the service could use an SLO to define the production readiness by its limited and general availability or other aspects of the service. The availability of a service typically has deadlines, and an SLO can assist in making a clear decision about the risk for a service.

Reviewer 2: W.r.t to this section:

"...production readiness by its limited and general availability or..."

limited availability is probably only understood by folks at Red Hat. IMO, this sentence doesn't lose anything by making it simply:

"Product managers responsible for the service could use an SLO to define the production readiness."

> Do we focus our time on making the service more reliable or adding a feature that customer X wanted?

Reviewer 3: There's ambiguous "we" again here that I'm not quite sure what to suggest for a less ambiguous reading.

> The answer to the above is your decision to make - SLO is the tool used to remove subjectivity about components of the decision. 

Reviewer 3: Who is "your" here?

> To learn more about the responsibilities across different personas in your organization, read our [article](./ADRs/RH/SIG-SRE/ADR-00006%20SLO%20RACI%20Chart.md) on the RACI model

Reviewer 3: ```suggestion
To learn more about the responsibilities across different personas in your organization, read our [article](./ADRs/RH/SIG-SRE/ADR-00006%20SLO%20RACI%20Chart.md) on the RACI model.
```

## Case 2 - Existing Service 

For this scenario, imagine we created our SLIs for an existing Authentication service in production and defined an SLO. Suddenly, one day, we get paged that our SLO exceeded our target level.

> After a quick investigation, we notice that our 3rd party auth provider is down. Consequently, our auth service is in degraded mode. 

Reviewer 3: ```suggestion
For this scenario, imagine we created our SLIs for an existing Authentication service in production and defined an SLO. Suddenly, one day, we get paged that our SLO exceeded our target level.
After a quick investigation, we notice that the service's 3rd party auth provider is down. Consequently, our auth service is in degraded mode. 
```

Who is "we" here (in both usages)? It's a good opportunity to loop in the multiple personas (such as SRE) as the recipient of the alert.

> To learn more about the responsibilities across different personas in an organization, read our [article](./ADRs/RH/SIG-SRE/ADR-00006%20SLO%20RACI%20Chart.md) on the RACI model

Reviewer 3: ```suggestion
To learn more about the responsibilities across different personas in an organization, read our [article](./ADRs/RH/SIG-SRE/ADR-00006%20SLO%20RACI%20Chart.md) on the RACI model.
```

Missing `.`.

When creating an SLO, external dependencies should be considered, and mitigation strategies should be employed to handle service disruptions from those external dependencies. 
Once your organization is convinced of the benefits that SLOs bring across multiple teams, retrofitting this into a mature service has some cultural considerations. 

First, all stakeholders have to agree on the outcomes of missing an SLO [burning the error budget](./ADRs/RH/SIG-SRE/ADR-00007-Error-Budget-Policy.md). 
This decision may take time for engineering teams to digest the idea of slowing feature work and focusing on reliability issues. 
The goal here should be the balance of direct impact on current processes vs introducing new processes. 

There are two main paths this could be navigated, and each should consider if there is pre-existing data that provides a baseline of current service levels. 
* Set a high SLO to start with and work towards it. What is high? 99.9%? This decision depends on if you can approximate the current service levels being delivered. 
* Set a low SLO to be more tight interally and catch potential issues earlier. However, tight SLOs lead to higher alerting frequency. 

# Tips

> * Reflect on customer experience: Customer satisfaction drives the sustainability of the business, so make sure to focus on critical paths that make customers happy.

Reviewer 2: Number one tip is "having good SLIs based on what is important to customers". :)

Author: I believe this is referred multiple times to our SLI guide and wanted to focus only on SLO stuff and let the reader follow the "Picking good SLIs" guide for more SLI tips

* Define Expectations vs Reality SLOs: Distinct aspirational SLO target levels from those you can acually achieve.
* Start somewhere: Don't wait until you find the best target levels to create an SLO. 
* Iterate often: You need continuously re-evaluate your reliability targets and update the SLOs based on more realistic target goals.
* Never forget external dependencies: Monitoring external dependencies and considering them while crafting SLOs is important as they have a direct impact on your reliability targets.

> To learn more about the SLO lifecycle check our [article](./ADRs/RH/SIG-SRE/ADR-00005%20SLO%20Lifecycle.md) to give you a glimpse of our approach.  

Reviewer 3: ```suggestion
To learn more about the SLO lifecycle check our [article](./ADRs/RH/SIG-SRE/ADR-00005%20SLO%20Lifecycle.md).
```

We're certainly putting all our effort into these docs, so there's not a "glimpse" per se; it's all there with nothing held back.
